diff --git a/iod/src/MachineInstance.cpp b/iod/src/MachineInstance.cpp
index 98319af..a8cc9d4 100644
--- a/iod/src/MachineInstance.cpp
+++ b/iod/src/MachineInstance.cpp
@@ -161,6 +161,11 @@ bool ConditionHandler::check(MachineInstance *machine) {
 			std::cerr << machine->getName() << " error: flag " << flag_name << " not found\n";
 		else {
 			if (condition(machine)) {
+				if (tracing() && machine->isTraceable()) {
+					machine->resetTemporaryStringStream();
+					machine->ss << machine->getCurrentStateString() << " " << *condition.predicate;
+					machine->setValue("TRACE", machine->ss.str());
+				}
 				const State *on = flag->state_machine->findState("on");
 				if (!flag->isActive()) flag->setState(*on);
 				else {
@@ -170,6 +175,11 @@ bool ConditionHandler::check(MachineInstance *machine) {
 				}
 			}
 			else {
+				if (tracing() && machine->isTraceable()) {
+					machine->resetTemporaryStringStream();
+					machine->ss << machine->getCurrentStateString() << " " << *condition.predicate;
+					machine->setValue("TRACE", machine->ss.str());
+				}
 				const State *off = flag->state_machine->findState("off");
 				if (!flag->isActive()) flag->setState(*off);
 				else {
