diff --git a/iod/src/MachineInstance.cpp b/iod/src/MachineInstance.cpp
index 1e00653..36c4038 100644
--- a/iod/src/MachineInstance.cpp
+++ b/iod/src/MachineInstance.cpp
@@ -3024,6 +3024,14 @@ void MachineInstance::enable() {
 		}
 		else
 			locals[i].machine->enable();
+#if 1
+		{
+			std::string msgstr(_name);
+			msgstr += "_enabled";
+			Message *msg = new Message(msgstr.c_str());
+			sendMessageToReceiver(msg, this, false);
+		}
+#endif
 	}
 	if (_type == "LIST") {
 		// enabling a list enables the members after sorting them into dependency order
@@ -3082,6 +3090,14 @@ void MachineInstance::setDebug(bool which) {
 }
 
 void MachineInstance::disable() { 
+#if 0
+	{
+		std::string msgstr(_name);
+		msgstr += "_disabled";
+		Message *msg = new Message(msgstr.c_str());
+		sendMessageToReceiver(msg, this, false);
+	}
+#endif
 	is_enabled = false;
 	if (locked) locked = 0;
 	if (io_interface) {
